### ISSUE TITLE
Missing change detection since `@rx-angular/cdk@1.0.0-beta.3`

### ISSUE THREAD

**Issue by twittwer** [rx-angular/rx-angular#1385]
# Description

Since our Angular 14 update which included upgrading `@rx-angular/cdk` from `1.0.0-beta.2` to `1.0.0-beta.3` we got some problems with missing change detections.

We have an application with a splitted view with some `*rxLet` directives. With the update we encounter the problem, that changes triggered in the main view aren't rendered in the sidebar, only after a clicking some UI element in the sidebar it gets rendered.

We found 2 options to fix the problem:
1) Replace `*rxLet="vm$; let vm"` with `*ngIf="vm$ | async as vm"`
2) Downgrading `@rx-angular/cdk` to `1.0.0-beta.2`

Maybe it is related to the coalescing optimizations in the [latest version](https://github.com/rx-angular/rx-angular/blob/main/libs/cdk/CHANGELOG.md#100-beta3-2022-06-07).

# Environment

```
Angular CLI: 14.1.1
Node: 16.13.2
OS: Windows 10

Angular: 14.1.1
Ivy Workspace: Yes

"@rx-angular/cdk": "1.0.0-beta.3",
"@rx-angular/state": "1.6.1",
"@rx-angular/template": "1.0.0-beta.33",
```

**Comment by hoebbelsB** [comment]
hey @twittwer thx for the issue!
Is there any way you can give us more information about the code? A tiny stackblitz example providing some snippets about how `vm$` is constructed would be amazing!

**Comment by twittwer** [comment]
Thx for your quick response

I cannot promise but will try to reproduce it in a small environment

**Comment by twittwer** [comment]
It seems to be a bit more complicated to reproduce the behaviour in minimalistic setup.
I cannot continue for the rest of the week, but here is my first try to show the setup: https://stackblitz.com/edit/angular-ivy-jfi7f6
Maybe I have to add some of the NgRx setup to introduce some more asynchronicity.

**Comment by hoebbelsB** [comment]
Can we observe some weird behavior in the current stackblitz example? 

> Maybe I have to add some of the NgRx setup to introduce some more asynchronicity.

NgRx shouldn't be async as far as I can tell

**Comment by twittwer** [comment]
No, the Stackblitz seems to behave expected.
But we experience the problem even with the new RC versions.

**Comment by twittwer** [comment]
@hoebbelsB Now I can provide a ~minimal reproduction:  
https://github.com/twittwer/repro-rx-angular-1385

After checkout the problem can be enabled/disabled by switching between the `@rx-angular/cdk` versions `1.0.0-beta.2` (without problem) and `@rx-angular/cdk@1.0.0-rc.1` (with problem).

**Comment by exequiel09** [comment]
I came across this weird behavior as well. I'm not okay with calling the `ChangeDetectorRef.markForCheck()`. This happened to me when I upgraded to Angular v14. In Angular v13, it's working fine.

**Comment by twittwer** [comment]
I stripped out NgRx & the Angular Router for a simpler reproduction scenario:
https://github.com/twittwer/repro-rx-angular-1385

**Comment by hoebbelsB** [comment]
@twittwer is there a specific reason why you are not using the very latest versions of rx-angular? just curious

**Comment by hoebbelsB** [comment]
@exequiel09 do I understand correctly this behavioral change is coming from angular 14, not from rx-angular? @twittwer did u change the angular version of your project or did u just change rx-angular and the error was introduced clearly with that version change? 
I'm trying to wrap my head around this, but the example is a bit complex. I guess it has something to do with the way how the `ngTemplateOutlet` is used in your case.

Probably a [`cdkPortal`](https://material.angular.io/cdk/portal/api) would help because you can configure the `ViewContainerRef`.

What I can observe is that `rxLet` is actually updating the state of its embeddedView, you can verify this if you add a binding for ` {{ vm.activeTodo.description }}` just next to the `*appSidebar` usage.
The value inside of `*appSidebar` is not updated, because it actually is another templateRef, injected into another `ViewContainerRef`. The main problem is: the other `ViewContainerRef` is not a child of the current rxLet, but a sibling of the component hosting the rxLet.

**Comment by hoebbelsB** [comment]
Compared the flamecharts against each other and it looks like with `beta.2` there was an additional NgZone call inside of the scheduler queue which is missing in `beta.3`. So the change should be visible somewhere in the scheduler implementation. Very interesting :) 

![no-ngzone-tick](https://user-images.githubusercontent.com/4904455/203434698-9b84131e-1175-43f2-a7a7-ae2e91060011.png)

**Comment by hoebbelsB** [comment]
It was probably this PR https://github.com/rx-angular/rx-angular/pull/946 introducing the changed behavior. Didn't check 100%, but at least I know we've changed NgZone and scheduler behavior there.

**Comment by exequiel09** [comment]
@hoebbelsB I'll be adding my reproduction soon once I get to work on it.

**Comment by twittwer** [comment]
@hoebbelsB At first, thank you for looking into it 🙏 

> @twittwer is there a specific reason why you are not using the very latest versions of rx-angular? just curious

The reason is this change of the change detection which breaks our application. That's why we are currently stuck on RxAngular beta.2 and Angular < 14.2 because of #1386.

>  @twittwer did u change the angular version of your project or did u just change rx-angular and the error was introduced clearly with that version change?

As shown in the reproduction, the behaviour changes by changing the RxAngular version (more precise `@rx-angular/cdk`) only.

> It was probably this PR https://github.com/rx-angular/rx-angular/pull/946 introducing the changed behavior. Didn't check 100%, but at least I know we've changed NgZone and scheduler behavior there.

When I was looking at the [changelog for beta.3](https://github.com/rx-angular/rx-angular/blob/main/libs/cdk/CHANGELOG.md#100-beta3-2022-06-07) the following 3 were looking suspicious to me:
- cdk: coalesce and optimize parent notification (https://github.com/rx-angular/rx-angular/issues/1262) ([9d1d099](https://github.com/rx-angular/rx-angular/commit/9d1d099608ffe848af207475a51f3788b94ca8bc))
- cdk: improve ngZone aware scheduler poc ([f6bab88](https://github.com/rx-angular/rx-angular/commit/f6bab88e096a6bbce225764480921f51e4024ae0))
- coalesce ngZone PoC ([668c28a](https://github.com/rx-angular/rx-angular/commit/668c28ad0fb949666c58b27468188d759cfaf5c0))

**Comment by hoebbelsB** [comment]
hey @twittwer invested a bit more time yesterday and now i'm pretty sure what is happening. Let me first say thank you for finding this, it was def. unexpected and looks like a bug in our scheduler implementation.

Let me try to explain what is going on here. Before https://github.com/rx-angular/rx-angular/commit/668c28ad0fb949666c58b27468188d759cfaf5c0 got merged, every single task executed in the scheduler queue was wrapped with `NgZone#run`, which led to performance issues as documented in the PR https://github.com/rx-angular/rx-angular/pull/946

The solution was to have a single `NgZone#run` executing all tasks in a scheduler tick. That way we reduced the amounts of NgZone calls and sideeffects coming from it.

pseudo code of the current scheduler flushWork fn:

```ts
performingWork = true;
NgZone.run(() => {
  while ( !hitDeadline && hasTask) {
     // run every task as long as we have time for it
     currentTask.execute();
   }
})
// => here the error occurs.
// After NgZone.run, NgZone executes code internally - without us knowing...
// the code can lead to other rxLet directives adding new tasks to the queue
// but, because `performingWork` is still true, the scheduler won't pick it up in the same tick and won't schedule a new
// tick
perfromingWork = false;
```

I will put together a PR with a fix for it :)

**Comment by exequiel09** [comment]
I can't put a decent minimal reproduction but @hoebbelsB I can help testing your fix in the codebase I'm working on. Hopefully that fixes my issue as well. 🙏🏻

**Comment by hoebbelsB** [comment]
> > exequiel09 do I understand correctly this behavioral change is coming from angular 14, not from rx-angular
> 
> @hoebbelsB I'm using the latest versions of rx-angular with v13. When I upgraded to v14, rxLet / rxFor in some places don't trigger initial render. When i switch back to v13, it works fine. I can try @twittwer findings to downgrade the cdk package and see if the same scenario happens when on v14 with downgraded cdk version.

maybe this is even yet another issue

**Comment by hoebbelsB** [comment]
> I can't put a decent minimal reproduction for the scenario I'm having with rxLet and rxFor but @hoebbelsB I can help testing your fix in the codebase I'm working on. Hopefully that fixes my issue as well. 🙏🏻

you might want to help me implementing a unit test for the current bug. The fix is in place here: https://github.com/rx-angular/rx-angular/pull/1478

**Comment by hoebbelsB** [comment]
@twittwer https://github.com/rx-angular/rx-angular/pull/1478 this PR should fix the issue. Can you please confirm this is solving your issue?

**Comment by hoebbelsB** [comment]
@twittwer exactly. don't forget to turn off ng build cache and remove the .cache folder

**Comment by twittwer** [comment]
Cool, your PR fixes the problem 🎉

**Comment by exequiel09** [comment]
@hoebbelsB not sure if totally related. It might be a different issue? not sure but I have 3 scenarios using Angular v14 and the latest @rx-angular/cdk in NPM.

1. NOT WORKING - Component that instantiates the Bottom Sheet has Change Detection Strategy set to OnPush. And the button that triggers to show the bottom sheet has the event listener running outside the Angular's zone via `@rx-angular`'s `unpatch` directive. But before the bottom sheet is created, the execution is returned inside NgZone via enterNgZone function of `ngx-rxjs-zone-scheduler`. (https://stackblitz.com/edit/angular-ivy-k32w43?file=src/app/app.component.ts).
2. WORKING - Component that instantiates the Bottom Sheet is not set. And the button that triggers to show the bottom sheet has the event listener running outside the Angular's zone via `@rx-angular`'s `unpatch` directive. But before the bottom sheet is created, the execution is returned inside NgZone via enterNgZone function of `ngx-rxjs-zone-scheduler`. (https://stackblitz.com/edit/angular-ivy-7cfpb1?file=src/app/app.component.ts)
3. WORKING - Component that instantiates the Bottom Sheet has Change Detection Strategy set to OnPush. And the button that triggers to show the bottom sheet has the event listener running outside the Angular's zone via `@rx-angular`'s `unpatch` directive. But before the bottom sheet is created, the execution is returned inside NgZone by wrapping the MatBottomSheet.open in `NgZone.run()` call. (https://stackblitz.com/edit/angular-ivy-ajj7dr?file=src%2Fapp%2Fapp.component.ts)

This stackblitz link https://stackblitz.com/edit/angular-ivy-inttfu?file=src%2Fapp%2Fapp.component.ts uses Angular v13 in scenario 1 and it still works

**Comment by exequiel09** [comment]
> NOT WORKING - Component that instantiates the Bottom Sheet has Change Detection Strategy set to OnPush. And the button that triggers to show the bottom sheet has the event listener running outside the Angular's zone via @rx-angular's unpatch directive. But before the bottom sheet is created, the execution is returned inside NgZone via enterNgZone function of ngx-rxjs-zone-scheduler. (https://stackblitz.com/edit/angular-ivy-k32w43?file=src/app/app.component.ts).

Weird behavior for scenario 1. When I use the AppComponent's `ViewContainerRef` change happens but not when using the read ViewContainerRef from a `ng-container`. See: https://stackblitz.com/edit/angular-ivy-kahenx?file=src%2Fapp%2Fapp.component.ts
